File: questions.java



Recover a bst
https://leetcode.com/problems/recover-binary-search-tree/
class Solution {
    
    TreeNode a = null , b = null,prev = null;
    public boolean recover(TreeNode root)
    {
        if(root == null) return false;
        
        if(recover(root.left)) return true;
        
        if(prev!=null && prev.val >  root.val)
        {
            b = root;
            if(a == null) a = prev;
            else return true;
        }
        
        prev = root;
        if(recover(root.right)) return true;
        
        return false;
    }
    
    
    public void recoverTree(TreeNode root) {
      
        recover(root);
        
        if(a!=null)
        {
            int temp = a.val;
            a.val = b.val;
            b.val = temp;
        }
    }
}




// 
to check if the tree is same or not!


class Solution {
    public boolean isSameTree(TreeNode p, TreeNode q) {
     
        
        if(p == null && q == null) return true;
        else if (p== null || q == null) return false;
        
        
        if(p.val!=q.val) return false;
        
        
    
        return isSameTree(p.left,q.left) && isSameTree(p.right,q.right);
    }
}


// to check if the trees are symettric or not

public boolean isSymmetric_(TreeNode root1,TreeNode root2){
        
        if(root1 == null && root2 == null) return true;
        else if(root1 == null || root2 == null || root1.val!=root2.val) return false;
        return isSymmetric_(root1.left,root2.right) && isSymmetric_(root1.right,root2.left);
    }
    public boolean isSymmetric(TreeNode root) {
        if(root == null) return true;
        return isSymmetric_(root.left,root.right);
    }




// level order traversal of tree

public void solve(TreeNode root,  List<List<Integer>>  ans)
    {
        if(root == null) return;
        
        
        LinkedList<TreeNode> que  = new LinkedList<>();
        
        que.addLast(root);
        
        while(que.size()!=0){
            int size =que.size();
            ArrayList<Integer> res = new ArrayList<>();
            while(size-->0){
                TreeNode vtx = que.removeFirst();
                
                res.add(vtx.val);
                
                if(vtx.left!=null) que.addLast(vtx.left);
                 if(vtx.right!=null) que.addLast(vtx.right);
            }
            
            ans.add(res);
        }
    }
    public List<List<Integer>> levelOrder(TreeNode root) {
        
        List<List<Integer>>  ans = new ArrayList<>();
        solve(root,ans);
        
        
        return ans;
        
    }



right view of a bst
public void rightview(TreeNode root,ArrayList<Integer> ans)
    {
        LinkedList<TreeNode> qu = new LinkedList<>();
        qu.add(root);
        
        
        while(qu.size()!=0)
        {
            int size = qu.size();
            ans.add(qu.getFirst().val);
            while(size -- >0)
            {
                
                TreeNode vtx = qu.removeFirst();
                
                if(vtx.right!=null) qu.addLast(vtx.right);
                if(vtx.left != null) qu.addLast(vtx.left);
                
            }
        }
    }
    public List<Integer> rightSideView(TreeNode root) {
        if(root == null)return new ArrayList<>();
        
      ArrayList<Integer> ans = new ArrayList<>();
        
        rightview(root,ans);
        return ans;
        
    }

  post order traversal
  
   public class pair{
        TreeNode node = null;
        boolean leftdone = false;
        boolean rightdone = false;
        boolean selfdone = false;
        
        pair(TreeNode node ,  boolean selfdone, boolean leftdone , boolean rightdone){
            this.node = node;
            this.selfdone = selfdone;
            this.leftdone = leftdone;
            this.rightdone = rightdone;
        }
    }
    
    
    
    public List<Integer> postorderTraversal(TreeNode root) {
     if(root == null) return new ArrayList<>();
        ArrayList<Integer> ans = new ArrayList<>();
        
        Stack<pair> st = new Stack<>();
        st.push(new pair(root,false,false,false));
        while(st.size()!=0){
            pair rp = st.peek();
            
            if(!rp.leftdone){
                rp.leftdone = true;
                if(rp.node.left!=null) st.push(new pair(rp.node.left,false,false,false));
            }else if(!rp.rightdone){
                rp.rightdone = true;
                if(rp.node.right!=null) st.push(new pair(rp.node.right,false,false,false));
            }else if(!rp.selfdone){
                rp.selfdone = true;
                ans.add(rp.node.val);
            }else{
                st.pop();
            }
        }
        return ans;
    }

find the sum from root

public int NumSum(TreeNode root ,int val){
        if(root == null) return 0;
        
        val = val*10 + root.val;
        
        if(root.left == null && root.right == null) return val;
        
        
        return NumSum(root.left,val)+NumSum(root.right,val);
    }
    
    public int sumNumbers(TreeNode root) {
        return NumSum(root,0);
        // return a;
    }


preorder traversal

public class pair{
        TreeNode node = null;
        boolean selfdone = false;
        boolean leftdone = false;
        boolean rightdone = false;
        
        pair(TreeNode node,boolean selfdone,boolean leftdone,boolean rightdone){
            this.node = node;
            this.selfdone = selfdone;
            this.leftdone = leftdone;
            this.rightdone = rightdone;
        }
    }
    
    
    public List<Integer> preorderTraversal(TreeNode root) {
        if(root == null) return new ArrayList<>();
        ArrayList<Integer> ans = new ArrayList<>();
        Stack<pair>  st = new Stack<>();
        st.push(new pair(root,false,false,false));
        
        
        while(st.size()!=0){
            pair rp = st.peek();
            
            if(!rp.selfdone){
                rp.selfdone = true;
                ans.add(rp.node.val);
                // System.out.print(rp.node.val+" ");
            }else if(!rp.leftdone){
                rp.leftdone = true;
                if(rp.node.left!=null) st.push(new pair(rp.node.left,false,false,false));
            }else if(!rp.rightdone){
                rp.rightdone = true;
                if(rp.node.right!=null) st.push(new pair(rp.node.right,false,false,false));
            }else{
                st.pop();
            }
            
        }
        return ans;
    }

File: questions.cpp

Hi 
The differences between cpp and java are as follows:-
Java doesn't support default arguments like C++. 
Java does not support header files like C++. 
Java uses the import keyword to include different 

1.
Most imp question :
Tree Traveral
struct Node {
    int data;
    struct Node *left, *right;
    Node(int data)
    {
        this->data = data;
        left = right = NULL;
    }
};
 
/* Given a binary tree, print its nodes according to the
"bottom-up" postorder traversal. */
void printPostorder(struct Node* node)
{
    if (node == NULL)
        return;
 
    // first recur on left subtree
    printPostorder(node->left);
 
    // then recur on right subtree
    printPostorder(node->right);
 
    // now deal with the node
    cout << node->data << " ";
}
 
/* Given a binary tree, print its nodes in inorder*/
void printInorder(struct Node* node)
{
    if (node == NULL)
        return;
 
    /* first recur on left child */
    printInorder(node->left);
 
    /* then print the data of node */
    cout << node->data << " ";
 
    /* now recur on right child */
    printInorder(node->right);
}
 
/* Given a binary tree, print its nodes in preorder*/
void printPreorder(struct Node* node)
{
    if (node == NULL)
        return;
 
    /* first print data of node */
    cout << node->data << " ";
 
    /* then recur on left subtree */
    printPreorder(node->left);
 
    /* now recur on right subtree */
    printPreorder(node->right);
}
 
/* Driver program to test above functions*/
int main()
{
    struct Node* root = new Node(1);
    root->left = new Node(2);
    root->right = new Node(3);
    root->left->left = new Node(4);
    root->left->right = new Node(5);
 
    cout << "\nPreorder traversal of binary tree is \n";
    printPreorder(root);
 
    cout << "\nInorder traversal of binary tree is \n";
    printInorder(root);
 
    cout << "\nPostorder traversal of binary tree is \n";
    printPostorder(root);
 
    return 0;
}

/// path sum
class Solution {
public:
    bool helper(TreeNode* root, int targetSum) {
        if (!root) return false;
        if (!root->left && !root->right) {
            return targetSum - root->val == 0;
        }
        
        return helper(root->left, targetSum - root->val) || helper(root->right, targetSum - root->val);
    }
    bool hasPathSum(TreeNode* root, int targetSum) {        
        return helper(root, targetSum);
    }
};



// sum rot to leaf
class Solution {
public:
    int sol=0;
    int sumNumbers(TreeNode* root) {
        build(root,0);
        return sol;
    }
    void build(TreeNode* root,int n)
    {
        if(root)
        {
             n=n*10+root->val;
            if(root->left==NULL && root->right==NULL)
            {
                sol+=n;
                return ;
            }
            build(root->left,n);
            build(root->right,n);
        }
    }
};

class Solution {
public:
    int sol=0;
    int sumNumbers(TreeNode* root) {
        build(root,0);
        return sol;
    }
    void build(TreeNode* root,int n)
    {
        if(root)
        {
             n=n*10+root->val;
            if(root->left==NULL && root->right==NULL)
            {
                sol+=n;
                return ;
            }
            build(root->left,n);
            build(root->right,n);
        }
    }
};

//  C++ || DFS || BFS || DFS IS BETTER IN BOTH T.C AND S.C
class Solution {
public:
    //DFS 
   /* int ans=0;
    void solve(int h, TreeNode* root, int cur_depth){
        if(root==NULL) return;
        if(cur_depth==h && root->left==NULL && root->right==NULL) ans+=root->val;
        solve(h,root->left,cur_depth+1);
        solve(h,root->right,cur_depth+1);
     }
    
    
    int depth(TreeNode* root){
        if(root==NULL) return 0;
         int ld=depth(root->left);
          int rd=depth(root->right);
          return 1+max(ld,rd);
    }
    
    int deepestLeavesSum(TreeNode* root) {
       int max_depth= depth(root);
        solve(max_depth,root,1);
        return ans;
    }
     
//BFS
    /* 
	int deepestLeavesSum(TreeNode* root) {
         if(root==NULL) return 0;
          vector<int> temp;
          queue<TreeNode*> q;
          q.push(root);
         while(!q.empty()){
             temp.clear();           
             int sz=q.size();
             
             for(int i=0; i<sz; i++){
                 auto x=q.front();
                 q.pop();
                 temp.push_back(x->val);
                 if(x->left) q.push(x->left);
                 if(x->right) q.push(x->right);
             }
             
         }
             return accumulate(temp.begin(),temp.end(),0);
    }
    */
};


// C++ | Iterative | Level Order Traversal | BFS | TC : O(N) SC : O(N)
int deepestLeavesSum(TreeNode* root) {
		priority_queue<pair<int,int>> maxHeap;
        queue<TreeNode*> Q;
		
        Q.push(root);
        Q.push(new TreeNode(-1));
        int depth = 0;
		
        while(!Q.empty()){
            if(Q.front()->val == -1) {
                Q.push(new TreeNode(-1));
                Q.pop();
                depth++;
            }
            
            TreeNode *tempNode = Q.front();
            Q.pop();
            
            if(!tempNode->left && !tempNode->right && !(tempNode->val == -1))
                maxHeap.push(make_pair(depth,tempNode->val));
            
            if(tempNode->left)
                Q.push(tempNode->left);
            
            if(tempNode->right)
                Q.push(tempNode->right);
        }
        
        int sum = 0, maxDepth = maxHeap.top().first;
        while(!maxHeap.empty()){
            if(maxHeap.top().first != maxDepth)
                break;
            sum += maxHeap.top().second;
            maxHeap.pop();
        }
        
        return sum;
    }
